### PR TITLE
AWS wrapper for tracing

### PR DESCRIPTION
**Pull request by stroem**

### CODE REVIEW

File: v2/trace/aws/extract.go [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+			dd_tracer.Tag("request_id", lambdaCtx.AwsRequestID),
+			dd_tracer.Tag("resource_names", lambdacontext.FunctionName),
+		)
+	}
+
+	if parentSpanContext == nil {
+		return dd_tracer.StartSpanFromContext(ctx, operationName, spanOpts...)
+	}
+
+	spanOpts = append(spanOpts, dd_tracer.ChildOf(parentSpanContext))
```
Comment (lonnblad) [inline v2/trace/aws/extract.go:108]: maybe:
```
spanOpts = append(spanOpts, dd_tracer.ChildOf(parentSpanContext))
recordSpan := dd_tracer.StartSpan(operationName, spanOpts...)

return recordSpan, dd_tracer.ContextWithSpan(ctx, recordSpan)
```

**Comment by lonnblad** [comment]
Missing a docs.go explaining the package and how to use it as well.

File: v2/trace/aws/extract.go [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+				traceAttributes[key] = attr.Value
+			}
+		}
+	}
+
+	return traceAttributes
+}
+
+func startSpan(ctx context.Context, serviceName string, parentSpanContext dd_trace.SpanContext) (dd_tracer.Span, context.Context) {
+	operationName := "sns.handler"
```
[the remark this answers is not in the document]
Comment (stroem) [reply v2/trace/aws/extract.go:82]: Unsure of what. Any suggestions? "record.handler"?

File: v2/trace/aws/extract.go (hunk 1 above)
Comment (stroem) [reply v2/trace/aws/extract.go:108]: I really donät see any difference from what it is right now?

**Comment by stroem** [comment]
doc.go is added

**Review by lonnblad — APPROVED** [review]

File: v2/trace/aws/aws.go [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+
+	"github.com/aws/aws-sdk-go/aws/request"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/sns"
+	"github.com/aws/aws-sdk-go/service/sqs"
+)
+
+// WrapSession will wrap the AWS session and
+// injecting tracing headers into the outgoing requests for these operations:
+// - sns.Publish
```
Comment (lonnblad) [inline v2/trace/aws/aws.go:14]: we have "sns.Publish" in the doc, but then use "sns/Publish" later?

Comment (stroem) [reply v2/trace/aws/aws.go:14]: We can change the handler operation name to use dot to match datadog/oc trace

**Review by johdah — DISMISSED** [review]
Looks good, nice addition

File: v2/trace/aws/aws.go (hunk 3 above)
Comment (rickardenglund) [reply v2/trace/aws/aws.go:14]: I think we are using / in the applications

File: v2/trace/aws/doc.go [hunk 4]
```diff
@@ -0,0 +1,37 @@
+// Package awstrace contains aws session wrapper for tracing through SNS and SQS
```
Comment (rickardenglund) [inline v2/trace/aws/doc.go:1]: I quite like the idea of having examples in a separate package 
like put this in ./example/main.go and then not have is comments.

Then you get an compiler reminder to update docs when you are updating the code.

But maybe this is the go way?

File: v2/trace/aws/aws.go (hunk 3 above)
Comment (stroem) [reply v2/trace/aws/aws.go:14]: @rickardenglund Which applications?
I perfer dot in this case

Comment (rickardenglund) [reply v2/trace/aws/aws.go:14]: The applications being for example users and hierarchy. This results in the naming of spans right?

So I think this depends on whether you want it to blend in with the data dog stuff or the enlight stuff.

Either way I think "." is perfectly fine But it is nice to same in doc and code :)

File: v2/trace/aws/doc.go (hunk 4 above)
Comment (stroem) [reply v2/trace/aws/doc.go:1]: Good idea so the compiler catches all errors, but I still would like to have an short highlighted example in the doc.go

Comment (stroem) [reply v2/trace/aws/doc.go:1]: Better now?

File: v2/trace/aws/aws.go (hunk 3 above)
Comment (stroem) [reply v2/trace/aws/aws.go]: What about this solution, which highlight the functions it supports. Not which operations, which is not entirely the same

**Review by lonnblad — APPROVED** [review]

**Pull request merged**